python/ray/serve/_private/api.py

```diff
@@ -329,3 +334,52 @@ def serve_start(
         f'namespace "{SERVE_NAMESPACE}".'
     )
     return client
+
+
+def call_app_builder_with_args_if_necessary(
+    builder: Union[ClassNode, FunctionNode, FunctionType],
+    args: Dict[str, Any],
+) -> Union[ClassNode, FunctionNode]:
+    """Builds a Serve application from an application builder function.
+
+    If a pre-built application (ClassNode or FunctionNode) is passed, this is a no-op.
+
+    Else, we validate the signature of the builder, convert the args dictionary to
+    the user-annotated Pydantic model if provided, and call the builder function.
+
+    The output of the function is returned (must be a ClassNode or FunctionNode).
+    """
+    if isinstance(builder, (ClassNode, FunctionNode)):
+        if len(args) > 0:
+            raise ValueError(
+                "Arguments can only be passed to a builder function, "
+                "not an already built application."
+            )
+        return builder
+    elif not isinstance(builder, FunctionType):
+        raise TypeError(
+            "Expected a built Serve application or a builder function, "
+            f"but got: {type(builder)}."
+        )
+
+    # Check that the builder only takes a single argument.
+    # TODO(edoakes): we may want to loosen this to allow optional kwargs in the future.
+    signature = inspect.signature(builder)
+    if len(signature.parameters) != 1:
+        raise TypeError("Application builders should take exactly one parameter.")
```

Comment:
Should this describe what that one parameter is? Users may think that their function can take in arguments like `builder(arg1, arg2, arg3...)` instead of taking in a dict. My first thought when I saw this was, "the builder only allows one argument?"
```suggestion
        raise TypeError("Application builders should take exactly one parameter: a dictionary of arguments to be used in your builder function.")
```

Reply:
Also, curious: why don't we allow builders of the form `builder(arg1, arg2, arg3)` and instead force users to use `builder(args_dict)`?

Reply:
discussing second question on slack -- to your first, yes let me improve the error!